### ISSUE TITLE
conda environment corrupted when conda updates itself and other installation fails

### ISSUE THREAD

**Issue by aldanor** [conda/conda#1155]
I was just trying to install some recipe which failed at the pre-link stage and left the conda environment destroyed because it decided to update itself (just because it can)... that's not nice.
- conda unlinks itself and condaenv
- conda starts linking the package which fails at pre-link
- linking stage terminates, conda doesn't exist anymore in the base environment

**Comment by asmeurer** [comment]
Which package had the broken pre-link script?

**Comment by aldanor** [comment]
It's my own build recipe for Ruby :) 

That's quite funny actually (I know this is generally not an ideal way of doing this but serves as a good example of how things can go south):
```sh
    if [ -n "$DEFAULT_CONDA_ENV" ]; then
        echo "Can only install Ruby in the root conda environment."
        exit 1
    fi
    ROOT_PREFIX=`python -c "import os; from conda import config; \
        print os.path.realpath(config.root_dir)"`
    if [ "$PREFIX" != "$ROOT_PREFIX" ]; then
        echo "Can only install Ruby in the root conda environment."
        exit 1
    fi
```
So, conda happened to be unlinked while this pre-link was running because it wanted to update itself, so pre-link for ruby failed, which in turn made the WHOLE linking stage to fail, so conda never got to linking itself back in.

If a pre-link for a certain package fails, should the whole linking stage fail? If the package for which -pre-link failed has been previously unlinked, should conda attempt to link the old version back?

**Comment by asmeurer** [comment]
I see. I just wanted to make sure it wasn't one of our packages breaking. 

We should definitely make this roll back better.

**Comment by kalefranz** [comment]
Fixed with transactions in 4.3.
